index.js

```diff
@@ -79,17 +81,28 @@ function formatWithCursor(text, opts, addAlignmentSize) {
   const toStringResult = printDocToString(doc, opts);
   const str = toStringResult.formatted;
   const cursorOffsetResult = toStringResult.cursor;
+  const cursorNodeEnd = toStringResult.cursorNodeEnd;
   ensureAllCommentsPrinted(astComments);
   // Remove extra leading indentation as well as the added indentation after last newline
   if (addAlignmentSize > 0) {
     return { formatted: str.trim() + opts.newLine };
   }
 
   if (cursorOffset !== undefined) {
-    return {
-      formatted: str,
-      cursorOffset: cursorOffsetResult + cursorOffset
-    };
+    if (cursorOffsetResult + cursorOffset < cursorNodeEnd) {
+      return {
+        formatted: str,
+        cursorOffset: cursorOffsetResult + cursorOffset
+      };
+    } else {
+      return {
```

Comment:
Well, maybe. Before this PR, the cursor translation simply applied the offset from the beginning of the node the cursor was on. Now, it tries to do the same thing, unless that would move the cursor outside the node, in which case it applies the offset from the *end* of the node the cursor was on.

However, if the cursor doesn't escape the node after the beginning-offset is applied, it could be at a different position than if the end-offset is applied, so this logic gets a little complicated. Does that make sense?

---

Side note: at first, I thought you were talking about simply `return`ing, rather than putting it inside an `else` block, because the `if` already has a `return`. I'll open a separate PR for enforcing that lint rule.